pkg/compiler/validators/typing.go

```diff
@@ -130,6 +134,46 @@ func (t schemaTypeChecker) CastsFrom(upstreamType *flyte.LiteralType) bool {
 	return true
 }
 
+// Structured dataset are more complex types in the Flyte ecosystem. A structured dataset is considered castable in the following
+// cases.
+//
+//    1. The downstream structured dataset has no column types specified.  In such a case, it accepts all structured dataset input since it is
+//       generic.
+//
+//    2. The downstream structured dataset has a subset of the upstream columns and they match perfectly.
+//
+func (t structuredDatasetChecker) CastsFrom(upstreamType *flyte.LiteralType) bool {
+	// structured dataset are nullable
+	if isVoid(upstreamType) {
+		return true
+	}
+	structuredDatasetType := upstreamType.GetStructuredDatasetType()
```

Comment:
I think we want to allow for FlyteSchema as well actually.

Reply:
done

Reply:
I've tested it. check this [example](https://gist.github.com/pingsutw/20bba3ead52a9881ddb02b983efe1a5d)

![image](https://user-images.githubusercontent.com/37936015/148202777-69011f70-466b-456d-90f4-596eaf3002f7.png)